runtime/src/validate_block/implementation.rs

```diff
@@ -0,0 +1,230 @@
+	Block as BlockT, Header as HeaderT, Hash as HashT
+};
+use executive::ExecuteBlock;
+
+use substrate_trie::{MemoryDB, read_trie_value, delta_trie_root};
+
+use rstd::{slice, ptr, cmp, vec::Vec, boxed::Box, mem};
+
+use hash_db::HashDB;
+
+static mut STORAGE: Option<Box<StorageT>> = None;
+/// The message to use as expect message while accessing the `STORAGE`.
+const STORAGE_SET_EXPECT: &str =
+	"`STORAGE` needs to be set before calling this function.";
+const STORAGE_ROOT_LEN: usize = 32;
+
+/// Extract the hashing algorithm type from the given block type.
+type HashingOf<B> = <<B as BlockT>::Header as HeaderT>::Hashing;
+/// Extract the hash type from the given block type.
+type HashOf<B> = <B as BlockT>::Hash;
+
+/// Abstract the storage into a trait without `Block` generic.
+trait StorageT {
+	/// Retrieve the value for the given key.
+	fn get(&self, key: &[u8]) -> Option<Vec<u8>>;
+
+	/// Insert the given key and value.
+	fn insert(&mut self, key: &[u8], value: &[u8]);
+
+	/// Remove key and value.
+	fn remove(&mut self, key: &[u8]);
+
+	/// Calculate the storage root.
+	fn storage_root(&mut self) -> [u8; STORAGE_ROOT_LEN];
+}
+
+/// Validate a given parachain block on a validator.
+#[cfg(not(feature = "std"))]
+#[doc(hidden)]
+pub fn validate_block<B: BlockT, E: ExecuteBlock<B>>(
+	mut block_data: &[u8],
+) {
+	use codec::Decode;
+
+	let block_data = crate::ParachainBlockData::<B>::decode(&mut block_data)
+		.expect("Could not decode parachain block.");
+	// TODO: Add `PolkadotInherent`.
+	let block = B::new(block_data.header, block_data.extrinsics);
+	let storage = Storage::<B>::new(
+		block_data.witness_data,
+		block_data.witness_data_storage_root,
+	).expect("Witness data and storage root always match; qed");
+
+	let _guard = unsafe {
+		STORAGE = Some(Box::new(storage));
+		(
+			// Replace storage calls with our own implementations
+			rio::ext_get_allocated_storage.replace_implementation(ext_get_allocated_storage),
+			rio::ext_get_storage_into.replace_implementation(ext_get_storage_into),
+			rio::ext_set_storage.replace_implementation(ext_set_storage),
+			rio::ext_exists_storage.replace_implementation(ext_exists_storage),
+			rio::ext_clear_storage.replace_implementation(ext_clear_storage),
+			rio::ext_storage_root.replace_implementation(ext_storage_root),
+		)
+	};
+
+	E::execute_block(block);
+}
+
+/// The storage implementation used when validating a block.
+struct Storage<B: BlockT> {
+	witness_data: MemoryDB<<HashingOf<B> as HashT>::Hasher>,
+	overlay: hashbrown::HashMap<Vec<u8>, Option<Vec<u8>>>,
+	storage_root: HashOf<B>,
+}
+
+impl<B: BlockT> Storage<B> {
+	/// Initialize from the given witness data and storage root.
+	///
+	/// Returns an error if given storage root was not found in the witness data.
+	fn new(
+		data: WitnessData,
+		storage_root: HashOf<B>
+	) -> Result<Self, &'static str> {
+		let mut db = MemoryDB::default();
+		data.into_iter().for_each(|i| { db.insert(&[], &i); });
```

Comment:
what does the `&[]` mean here?

Reply:
I don't really know. I copied this code from substrate. I think that `&[]` is the prefix. Maybe @cheme can help us here?

Reply:
That is the prefix for the MemoryDB, I believe here if MemoryDB is used from substrate trie crate, there is no key transformation with the prefix so it is safe to put nothing (a variant called PrefixedMemoryDB does insert the prefix before the key when inserting/querying the keyvaluedb to avoid collision in trie with unhashed value).

Reply:
hm, maybe we need this when using a chain with child tries

Reply:
there is the key conflict with child trie, but this way of doing things cannot solve it (for two identical child trie there is the same prefix). I was also wondering about exposing this key transformation trait to insert the prefix with a specialized implementation (https://github.com/paritytech/trie/pull/12 last comment), but using a hashdb specific implementation would achieve the same thing.

Reply:
I assume that we then need to adapt [this](https://github.com/paritytech/substrate/blob/master/core/state-machine/src/proving_backend.rs#L104)? This is basically the functionality I use to generate the proof. I assume we would need to extend the proof with the prefix somehow?

Reply:
If you use child trie I think you need https://github.com/paritytech/substrate/pull/2209 (I should split it in two has I believe the api change would not be merged or at least requires some consensus but the proof side of the pr seems ok).
But otherwhise the prefix thing should only apply to avoid reference count in db, so the Recorder of trie does not use the prefix so it should not matter (that is the good thing about the way Arkadiy solves the issue).